libyul/backends/wasm/BinaryTransform.cpp

```diff
@@ -0,0 +1,549 @@
+using namespace dev;
+using namespace yul::wasm;
+
+namespace
+{
+
+enum class Section: uint8_t
+{
+	CUSTOM = 0x00,
+	TYPE = 0x01,
+	IMPORT = 0x02,
+	FUNCTION = 0x03,
+	MEMORY = 0x05,
+	GLOBAL = 0x06,
+	EXPORT = 0x07,
+	CODE = 0x0a
+};
+
+enum class ValueType: uint8_t
+{
+	Void = 0x40,
+	Function = 0x60,
+	I64 = 0x7e,
+	I32 = 0x7f
+};
+
+enum class Export: uint8_t
+{
+	Function = 0x0,
+	Memory = 0x2
+};
+
+enum class Opcode: uint8_t
+{
+	Unreachable = 0x00,
+	Nop = 0x01,
+	Block = 0x02,
+	Loop = 0x03,
+	If = 0x04,
+	Else = 0x05,
+	Try = 0x06,
+	Catch = 0x07,
+	Throw = 0x08,
+	Rethrow = 0x09,
+	BrOnExn = 0x0a,
+	End = 0x0b,
+	Br = 0x0c,
+	BrIf = 0x0d,
+	BrTable = 0x0e,
+	Return = 0x0f,
+	Call = 0x10,
+	CallIndirect = 0x11,
+	ReturnCall = 0x12,
+	ReturnCallIndirect = 0x13,
+	Drop = 0x1a,
+	Select = 0x1b,
+	LocalGet = 0x20,
+	LocalSet = 0x21,
+	LocalTee = 0x22,
+	GlobalGet = 0x23,
+	GlobalSet = 0x24,
+	I32Const = 0x41,
+	I64Const = 0x42,
+};
+
+static std::map<string, uint8_t> const builtins = {
+	{"i32.load", 0x28},
+	{"i64.load", 0x29},
+	{"i32.load8_s", 0x2c},
+	{"i32.load8_u", 0x2d},
+	{"i32.load16_s", 0x2e},
+	{"i32.load16_u", 0x2f},
+	{"i64.load8_s", 0x30},
+	{"i64.load8_u", 0x31},
+	{"i64.load16_s", 0x32},
+	{"i64.load16_u", 0x33},
+	{"i64.load32_s", 0x34},
+	{"i64.load32_u", 0x35},
+	{"i32.store", 0x36},
+	{"i64.store", 0x37},
+	{"i32.store8", 0x3a},
+	{"i32.store16", 0x3b},
+	{"i64.store8", 0x3c},
+	{"i64.store16", 0x3d},
+	{"i64.store32", 0x3e},
+	{"memory.size", 0x3f},
+	{"memory.grow", 0x40},
+	{"i32.eqz", 0x45},
+	{"i32.eq", 0x46},
+	{"i32.ne", 0x47},
+	{"i32.lt_s", 0x48},
+	{"i32.lt_u", 0x49},
+	{"i32.gt_s", 0x4a},
+	{"i32.gt_u", 0x4b},
+	{"i32.le_s", 0x4c},
+	{"i32.le_u", 0x4d},
+	{"i32.ge_s", 0x4e},
+	{"i32.ge_u", 0x4f},
+	{"i64.eqz", 0x50},
+	{"i64.eq", 0x51},
+	{"i64.ne", 0x52},
+	{"i64.lt_s", 0x53},
+	{"i64.lt_u", 0x54},
+	{"i64.gt_s", 0x55},
+	{"i64.gt_u", 0x56},
+	{"i64.le_s", 0x57},
+	{"i64.le_u", 0x58},
+	{"i64.ge_s", 0x59},
+	{"i64.ge_u", 0x5a},
+	{"i32.clz", 0x67},
+	{"i32.ctz", 0x68},
+	{"i32.popcnt", 0x69},
+	{"i32.add", 0x6a},
+	{"i32.sub", 0x6b},
+	{"i32.mul", 0x6c},
+	{"i32.div_s", 0x6d},
+	{"i32.div_u", 0x6e},
+	{"i32.rem_s", 0x6f},
+	{"i32.rem_u", 0x70},
+	{"i32.and", 0x71},
+	{"i32.or", 0x72},
+	{"i32.xor", 0x73},
+	{"i32.shl", 0x74},
+	{"i32.shr_s", 0x75},
+	{"i32.shr_u", 0x76},
+	{"i32.rotl", 0x77},
+	{"i32.rotr", 0x78},
+	{"i64.clz", 0x79},
+	{"i64.ctz", 0x7a},
+	{"i64.popcnt", 0x7b},
+	{"i64.add", 0x7c},
+	{"i64.sub", 0x7d},
+	{"i64.mul", 0x7e},
+	{"i64.div_s", 0x7f},
+	{"i64.div_u", 0x80},
+	{"i64.rem_s", 0x81},
+	{"i64.rem_u", 0x82},
+	{"i64.and", 0x83},
+	{"i64.or", 0x84},
+	{"i64.xor", 0x85},
+	{"i64.shl", 0x86},
+	{"i64.shr_s", 0x87},
+	{"i64.shr_u", 0x88},
+	{"i64.rotl", 0x89},
+	{"i64.rotr", 0x8a},
+	{"i32.wrap_i64", 0xa7},
+	{"i64.extend_i32_s", 0xac},
+	{"i64.extend_i32_u", 0xad},
+};
+
+bytes opcode(Opcode _o)
+{
+	return bytes(1, uint8_t(_o));
+}
+
+bytes lebEncode(uint64_t _n)
+{
+	bytes encoded;
```

Comment:
Do we need explicit initialisation here?

Reply:
No, that is only needed for POD types.